zcash_client_sqlite/src/wallet.rs

```diff
@@ -166,35 +267,115 @@ pub(crate) fn memo_repr(memo: Option<&MemoBytes>) -> Option<&[u8]> {
     })
 }
 
-pub(crate) fn get_max_account_id(
+// Returns the highest used account index for a given seed.
+pub(crate) fn max_zip32_account_index(
     conn: &rusqlite::Connection,
-) -> Result<Option<AccountId>, SqliteClientError> {
-    // This returns the most recently generated address.
-    conn.query_row_and_then("SELECT MAX(account) FROM accounts", [], |row| {
-        let account_id: Option<u32> = row.get(0)?;
-        account_id
-            .map(AccountId::try_from)
-            .transpose()
-            .map_err(|_| SqliteClientError::AccountIdOutOfRange)
+    seed_id: &HdSeedFingerprint,
+) -> Result<Option<zip32::AccountId>, SqliteClientError> {
+    conn.query_row_and_then(
+        "SELECT MAX(hd_account_index) FROM accounts WHERE hd_seed_fingerprint = :hd_seed",
+        [seed_id.as_bytes()],
+        |row| {
+            let account_id: Option<u32> = row.get(0)?;
+            account_id
+                .map(zip32::AccountId::try_from)
+                .transpose()
+                .map_err(|_| SqliteClientError::AccountIdOutOfRange)
+        },
+    )
+}
+
+struct AccountSqlValues<'a> {
+    account_type: u32,
+    hd_seed_fingerprint: Option<&'a [u8]>,
+    hd_account_index: Option<u32>,
+    ufvk: Option<String>,
+    uivk: String,
+}
+
+/// Returns (account_type, hd_seed_fingerprint, hd_account_index, ufvk, uivk) for a given account.  
+fn get_sql_values_for_account_parameters<'a, P: consensus::Parameters>(
+    account: &'a Account,
+    params: &P,
+) -> Result<AccountSqlValues<'a>, SqliteClientError> {
+    Ok(match account {
+        Account::Zip32(hdaccount) => AccountSqlValues {
+            account_type: AccountType::Zip32.into(),
+            hd_seed_fingerprint: Some(hdaccount.hd_seed_fingerprint().as_bytes()),
+            hd_account_index: Some(hdaccount.account_index().into()),
+            ufvk: Some(hdaccount.ufvk().encode(params)),
+            uivk: ufvk_to_uivk(hdaccount.ufvk(), params)?,
+        },
+        Account::Imported(ImportedAccount::Full(ufvk)) => AccountSqlValues {
+            account_type: AccountType::Imported.into(),
+            hd_seed_fingerprint: None,
+            hd_account_index: None,
+            ufvk: Some(ufvk.encode(params)),
+            uivk: ufvk_to_uivk(ufvk, params)?,
+        },
+        Account::Imported(ImportedAccount::Incoming(uivk)) => AccountSqlValues {
+            account_type: AccountType::Imported.into(),
+            hd_seed_fingerprint: None,
+            hd_account_index: None,
+            ufvk: None,
+            uivk: uivk.encode(&params.network_type()),
+        },
     })
 }
 
+pub(crate) fn ufvk_to_uivk<P: consensus::Parameters>(
```

Comment:
This method should either be documented as always producing the external UIVK, or take a `zip32::Scope` argument. Probably the former, given the apparent invariant that if `ufvk != NULL`, then `uivk` MUST be the external UIVK.